cpp/src/shuffler/shuffler.cpp

```diff
@@ -251,21 +257,24 @@ class Shuffler::Progress {
 
                 // Setup to receive the chunk into `in_transit_*`.
                 auto future =
-                    shuffler_.comm_->recv(src, gpu_data_tag, std::move(chunk.gpu_data));
+                    shuffler_.comm_->recv(src, gpu_data_tag, chunk.release_data_buffer());
                 RAPIDSMPF_EXPECTS(
-                    in_transit_futures_.insert({chunk.cid, std::move(future)}).second,
+                    in_transit_futures_.insert({chunk.chunk_id(), std::move(future)})
+                        .second,
                     "in transit future already exist"
                 );
                 RAPIDSMPF_EXPECTS(
-                    in_transit_chunks_.insert({chunk.cid, std::move(chunk)}).second,
+                    in_transit_chunks_.insert({chunk.chunk_id(), std::move(chunk)})
+                        .second,
                     "in transit chunk already exist"
                 );
                 shuffler_.statistics_->add_bytes_stat(
-                    "shuffle-payload-recv", chunk.gpu_data_size
+                    "shuffle-payload-recv", chunk.concat_data_size()
                 );
                 // Tell the source of the chunk that we are ready to receive it.
+                // TODO: all partition IDs in the chunk must map to the same key (rank).
```

Comment:
This is true by construction no? This chunk is extracted from `incoming_chunks_` which are populated by the deserialized chunks we received on line 206. The sender guaranteed on construction of that chunk that all the partitions mapped to our rank.

Reply:
Yes, that is true. So, it's guaranteed by the PostBox.